Applications RP Threat Model (#72)
radius-project/design-notes: architecture/2024-08-applications-rp-component-threat-model.md

# Radius Applications RP Component Threat Model

- **Author**: @nithyatsu

## Overview

This document provides a threat model for the Radius Applications RP component. It identifies potential security threats to this critical part of Radius and suggests possible mitigations. The document includes an analysis of the system, its assets, identified threats, and recommended security measures to protect the system.

> The Applications RP component is responsible for managing applications and their resources. It communicates with UCP, Deployment Engine and Controller components to achieve this. 

Reviewer: Does the Controller directly talk to the Applications-RP?

Reviewer 2: I think we could take this paragraph out. It duplicates content that's in the system description.

## Terms and Definitions

| Term                  | Definition                                                                                                                                                                                  |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| mTLS                  | Mutual Transport Layer Security (mTLS) allows two parties to authenticate each other during the initial connection of an SSL/TLS handshake.                                                 |
| UCP                 | Universal Control Plane for Radius                                                                                                                                                   |
## System Description

> Applications RP is a Radius service that acts as resource provider for application and its resources. The resources can be core resources like application or environment or container. They can also be a dapr resource, message queue or datastore. Applications RP lives in `radius-system` namespace in a kubernetes cluster. It is a client of Controller and UCP. It also receives requests from UCP for managing the above mentioned resources. 

Reviewer: We can list the available resource types we have here.

> Applications RP is a Radius service that acts as resource provider for application and its resources. The resources can be core resources like application or environment or container. They can also be a dapr resource, message queue or datastore. Applications RP lives in `radius-system` namespace in a kubernetes cluster. It is a client of Controller and UCP. It also receives requests from UCP for managing the above mentioned resources. 

Reviewer 2: I think we're missing a few points. Here's the big headlines I'd want to see covered. These are all **relevant** to the security discussion.

- Applications RP is a resource provider. Resource providers communicate over HTTP. 
- Applications RP needs a database, message queue, and secret store to function. These are all configurable components with multiple implementations.
- Users and clients do not interact with Applications RP directly. They communicate with UCP instead. 
- Applications RP manages resources in Kubernetes on the user's behalf. This may launch application code in the same Kubernetes cluster as Radius, or a different cluster.
- Applications RP can invoke *recipes* that execute Bicep or Terraform code.
- Applications RP has access to cloud credentials, and can manage cloud resources on the user's behalf.

### Architecture

Applications RP has four types of resource providers for managing various types of resources in an application. 

`Applications.Core` resource provider manages application, environment, container and gateways. 

`Applications.Dapr` resource provider manages all dapr resources that are deployed as part of application. 

`Applications.Datastore` resource provider manages datastore such as SQL database or Mongo DB that an application might be using. 

`Applications.Messaging` resources provider manages queues such as Rabbit MQ. 

> RecipeEngine is a key sub-component of Applications RP, which uses the above resource providers for provisioning infrastructure resources. It takes as input a recipe, written in bicep or terraform. It fetches Bicep recipe from any OCI complaint registry. It also fetches Terraform recipes available as public modules. It then requests UCP to help deploy the recipe. Also, if user creates a custom recipe, it requests Controller to first validate the recipe using recipe-webhook and then create it. 

Reviewer 2: > Also, if user creates a custom recipe, it requests Controller to first validate the recipe using recipe-webhook and then create it.

Can you say more? I'm not sure I get what this is saying.

Author: rewrote the document, these lines does not exist anymore.

> RecipeEngine is a key sub-component of Applications RP, which uses the above resource providers for provisioning infrastructure resources. It takes as input a recipe, written in bicep or terraform. It fetches Bicep recipe from any OCI complaint registry. It also fetches Terraform recipes available as public modules. It then requests UCP to help deploy the recipe. Also, if user creates a custom recipe, it requests Controller to first validate the recipe using recipe-webhook and then create it. 

Reviewer: Isn't the Recipe Engine inside the UCP?

Author: it is part of applications rp..

Users could provision application resources without using recipes too, in which case, Applications RP again works with Controller managing for Kubernetes resources and UCP for managing AWS and Azure resources. For deploying Terraform Recipe, it directly communicates with AWS and Azure.

> The Applications RP provisions some resources synchronously, whereas for other resources whose creation can be time consuming, it has workers that enable async operation. To faciliate Async operation, the Applications RP adds the incoming request to an in-memory queue. Workers dequeue requests from the queue and process them. 

Reviewer 2: Was this AI-generated?

Author: outdated content

> The Applications RP provisions some resources synchronously, whereas for other resources whose creation can be time consuming, it has workers that enable async operation. To faciliate Async operation, the Applications RP adds the incoming request to an in-memory queue. Workers dequeue requests from the queue and process them. 

Reviewer: UCP adds the messages for the async operations to the queue. Worker and Queue belong to UCP.

`/radius/deploy/Chart/templates/rp/configmaps.yaml` has the settings for all the providers used by Applications RP. 
```
storageProvider:
      provider: "apiserver"
      apiserver:
        context: ""
        namespace: "radius-system"
    queueProvider:
      provider: "apiserver"
      name: "radius"
      apiserver:
        context: ""
        namespace: "radius-system"
    metricsProvider:
      prometheus:
        enabled: true
        path: "/metrics"
        port: 9090
    profilerProvider:
      enabled: true
      port: 6060
    secretProvider:
      provider: kubernetes
    server:
      host: "0.0.0.0"
      port: 5443
    workerServer:
      maxOperationConcurrency: 10
      maxOperationRetryCount: 2
    ucp:
      kind: kubernetes
    logging:
      level: "info"
      json: true
    {{- if and .Values.global.zipkin .Values.global.zipkin.url }}
    tracerProvider:
      serviceName: "applications.core"
      zipkin: 
        url: {{ .Values.global.zipkin.url }}
    {{- end }}
```

#### Use of Cryptography

> 

Reviewer 2: Please fill this out.

#### Storage of secrets

> Applications RP stores secrets for rendering some resources. For instance TLS termination in gateways needs storing TLS cert and key. For these cases, Applications RP uses kubernetes secrets. 

Reviewer 2: Does this include all of the relevant cases?

Author: added several cases.

#### Data Serialization / Formats

> 

Reviewer 2: Well?

### Clients

**UCP** is the only client of Applications RP. It communicates with Applications RP for deploying various Radius Application resources. 

> All communications use mTLS.

Reviewer 2: Where did you find this info?

Author: removed the sentence.

## Trust Boundaries

We have a few different trust boundaries for the Applications RP component:

- **Kubernetes Cluster**: The overall environment where the Applications RP  component operates and serves clients.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads. 

The Applications RP component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCP and Controller also reside within the same namespace. Namespaces within Kubernetes can help set Role-Based Access Control (RBAC) policies.

Applications RP component communicates with Kube API server through Kube Controller for saving Radius resources and for queuing async operations. The API Server and Kube Controller run in the `kube-system` namespace within the cluster. 

## Assumptions

This threat model assumes that:

1. The Radius installation is not tampered with.
2. The Kubernetes cluster that Radius is installed on is not compromised.

> 3. It is the responsibility of the Kubernetes cluster to authenticate users. Administrators and users with sufficient privileges can perform their required tasks. Radius cannot prevent actions taken by an administrator.

Reviewer 2: Let's make sure to come back to the assumptions after we have some discussion.

## Data Flow

### Diagram

![Applications RP via Microsoft Threat Modeling Tool](2024-10-applications-rp-threat-model/apprp-dataflow.png)

Below are the key points associated with data flow:
1. Applications RP receives request to dpeloy resources from UCP and sends back appropriate response.
2. Applications RP communicates with Controller for managing kubernetes resources and validating custom recipes.
3. Applications RP requests UCP to deploy bicep recipes.
4.  Application RP (terraform provider) requests AWS/ Azure to deploy resources for terraform recipes.
5.  Application RP uses API server through Kubernetes Controller to save Radius resources and Async Operations ( APIServer is used as datastore and Queue).
6.  Application RP fetches recipes from OCI registries and public terraform modules.

> ### Threats

Reviewer 2: I'm going to hold off on reviewing threats until the first part of the document is done.

Author: we finished this.

# Radius Applications RP Component Threat Model

- **Author**: @nithyatsu

## Overview

This document provides a threat model for the Radius Applications RP component. It identifies potential security threats to this critical part of Radius and suggests possible mitigations. The document includes an analysis of the system, its assets, identified threats, and recommended security measures to protect the system.

The Applications RP component is responsible for managing applications and their resources. It communicates with UCP, Deployment Engine and Controller components to achieve this. 

## Terms and Definitions

| Term                  | Definition                                                                                                                                                                                  |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| RP                 | Resource Provider    |
| UCP                 | Universal Control Plane for Radius   |

## System Description

> Applications RP is a Radius component that acts as resource provider for application and its resources. It communicates over HTTP. The RP has a Datastore for storing Radius data, Message Queue for processing asynchronous request and a Secret Store for storing sensitive information such as certficates. All these are configurable components and support multiple implementations. Users and Clients cannot directly communicate with Applications RP. They instead communicate with UCP. UCP forwards relevant requests to Applications RP. Applications may have Kubernetes resources. Applications RP manages these resources on the user's behalf. This may launch user application's code on the same cluster as Radius, or a different cluster. It also has access to user's cloud credentials and manages user's cloud resources. Applications RP can invoke *recipes* which are bicep or terraform code. These recipes are used to deploy application infrastructure components like databases.  

Reviewer 2: - Applications-RP capabilities/responsibilities
- Dependencies (db, queue, secrets)
- Access patterns (HTTP, proxying from UCP, clients, and their interactions)

```suggestion
Applications RP is a Radius control-plane microservice that acts as resource provider for application and its resources. When users deploy an application using Radius, Applications RP manages the lifecycle of the created resources on the user's behalf. 

- Deploying an application may launch the user application's code on the same cluster as Radius, or a different cluster. 
- Deploying application may created resources in the cloud using Recipes. These *recipes* are bicep or terraform code, and are responsible for creating These  application infrastructure components like databases.
- As a result, Applications RP has access to the user's cloud credentials and can managed the user's cloud resources.

Resource providers (including Applications RP) communicate over HTTP and manage the lifecycle of resources. See the [architecture documentation](https://docs.radapp.io/concepts/technical/architecture/) for more context.  Users and clients cannot directly communicate with Applications RP. They instead communicate with UCP. UCP forwards relevant requests to Applications RP. 

The RP has a datastore for storing Radius data, message queue for processing asynchronous requests, and a secret Ssore for storing sensitive information such as certificates. All these are configurable components and support multiple implementations.  
```

### Architecture

> Application RP recieves HTTP requests from UCP. It does not interface directly with user/ cli. These requests have untrusted json payloads. RP validates these payloads before consuming them. 

Reviewer 2: Let's take this out, it duplicates information in the system description.

`Applications.Messaging` resources provider manages queues such as Rabbit MQ.

Applications RP has a key sub component `Recipe Engine` to execute `recipes`. 
`Recipes` are Bicep or Terraform code supplied by user that is used to deploy infrastructure components on Azure and AWS. The Bicep recipes are fetched from OCI compliant registries. Terraform recipes are public modules and fetched from internet too. 

In order to execute Terraform recipes, Applications RP installs latest Terraform. It also mounts an empty directory `/terraform` into Applications RP pod. It uses this directory for executing terraform recipes using the installed executable. The output resources generated from terraform module are converted to Radius output resources and stored in our datastore. 

In order to deploy bicep recipes, Applications RP sends a request to UCP, which in turn forwards it to Deployment Engine. 

The RP uses a queue to process requests asyncronously. Information about resources that are deployed / being deployed is stored in a datastore. 

> 

Reviewer 2: Let's talk about **how** we communicate and manage things in the cloud. 

- Recipes
- Things we managed directly (volumes, keyvault, managed identities)

Let's talk about **how** we communicate and manage things in Kubernetes. 

Example:

When a users deploys a gateway resource. Application RP uses its Kubernetes `ServiceAccount` to authenticate with the local Kubernetes cluster, and deploys `HTTPProxy` (Contour) resources to open an listener to the internet, and route traffic to the user's application. 

Example:

When a user deploys a container resource that needs a managed identity.... (please go find out)

![Applications RP](2024-10-applications-rp-threat-model/apprp.png) 

### Implementation Details

#### Use of Cryptography

1. **Computing the Hash for Terraform Backend Configuration**: [Link to code](https://github.com/radius-project/radius/blob/main/pkg/recipes/terraform/config/backends/kubernetes.go#L110).

   1. **Purpose**: The purpose of computing the hash is to compute a key for kubernetes secret which has the etrraform backend config
   2. **Library**: The library used to calculate the hash of the deployment configuration is the crypto library, which is one of the standard libraries of Go: [Link to library](https://pkg.go.dev/crypto@go1.23.1).
   3. **Type**: [SHA1](https://www.rfc-editor.org/rfc/rfc3174.html). Note: "SHA-1 is cryptographically broken and should not be used for secure applications." [Link to warning](https://pkg.go.dev/crypto/sha1@go1.23.1). This is used as an optimization for detecting changes, not as a security protection.

2. Applications RP also uses UCP's queue and store libraries. These libraries use Cryptography and are detailed in UCP threat model. [link to be added]

#### Storage of secrets

> Applications RP has access to sensitive information related to the application resource it manages as well as the cloud credentials it requires for managing cloud resources on Azure and AWS. Applications RP provides a Secret Store which can be used to store sensitive information such as TLS certificate and private keys. It uses kubernetes secrets to implement this secret store. 

Reviewer 2: Feedback - split this into two big ideas

- Access to cloud credentials (do we store them?)
- Data that Applications RP needs to store.

BIG TOPIC - let's add AZWI/IRSA to this discussion.

Example:

Applications RP can access the cloud using either static credentials or OIDC Federation (AzWI/IRSA). We recommend that users configure federation where possible as no secrets are required.

##### Managing secrets for datastores

> Applications RP service has a Datastore RP. This RP is the resource provider for datstores such as SQL database, Mongo DB and Redis Cache. As of today, sensitive information such as DB connection string, user/ password that is required to provision these resources is stored in plain text. The feature which enables datastores to use a secret store is in progress. 

Reviewer 2: It's really good that you identified a gap. Keep this in.

##### Access to cloud credentials and cloud

Applications RP requires AWS and Azure credentials for accessing and managing resources in cloud. It fetches credentials using UCP Secret Provider library. 
Credentials are not available for retrieval through API. Also, RP supports federated identities which enables robust security practices. 

Using these credentials, Applications RP can create other resources in AWS and Azure. The RP also can create managed identities for azure which will decide who can deploy and run user code.  

#### Access to cluster

> The RP can create kubernetes resources and manage them on behalf of the user. It can, for example, create a container based on the image provided by the user, which can in turn execute arbitrary code, and create other resources in the cluster where Radius is running, or on a different cluster. 

Reviewer 2: One question I have - **how** does Applications RP authenticate with Kubernetes?

Author: covered authentication using service-account

#### Exposing User Application to Internet

> The Radius RP can create ingress Kubernetes objects, which can expose a kubernetes service to internet. 

Reviewer 2: This is mostly true, we create Contour `HTTPProxy` objects.

#### Bicep Recipe execution

> `Recipes` are arbitrary Bicep or Terraform code which can provision infrastructure resources. In order to exceute a Bicep recipe, Applications RP's Recipe Engine first downloads the recipe from a OCI compliant registry over HTTP. Then it requests UCP to deploy the bicep. 

Reviewer 2: ```suggestion
Bicep recipes can provision arbitrary infrastructure resources in the cloud, and are provided by users. In order to exccute a Bicep recipe, Applications RP's Recipe Engine first downloads the recipe from a OCI compliant registry over HTTP. Then it requests UCP to deploy the bicep. 
```

Reviewer 2: How do we authenticate?

Author: covered auth.

#### Terraform Recipe execution

> Terraform recipes are download from internet too, since they are available as public modules. These recipes are downloaded onto an empty directory `/terraform` which is mounted into the applications RP pod. Then the installed terraform executable executes these recipes in the current directory. Terraform communicates with AWS and Azure as needed to deploy resources.

Reviewer 2: We support private modules as well as different kinds of authentication.

Terraform downloads providers from the internet and executes them inside Applications RP's pod. This means that they have access to the network, filesystem, and environment variables, as well as the running memory of applications rp.

We use custom parsers to parse Radius-related resource IDs and do not use any other custom parsers. The custom parser that parses Radius resource IDs has its own security mechanisms that don't accept anything other than a Radius resource ID. 

We also use json for data interchange between Applications RP and other services. We use golang standard libraries to parse json. 

## Clients

**UCP** UCP is the client of Applications RP.  It forwards all requests related to Applications and several of Application resources deployment to Applications RP.

## Trust Boundaries

We have a few different trust boundaries for the Controller component:

- **Kubernetes Cluster**: The overall environment where the Applications RP  operates and receives requests from the UCP.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads.

The Applications RP component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCPD also resides within the same namespace.

The Kubernetes API Server, with which Applications RP interacts, runs in the `kube-system` namespace within the cluster.

> 

Reviewer 2: What trust model do we apply to requests coming in to Applications RP?

- We assume that the actor is an authorized user, and has authorization make the request. 

How we do we enforce this?

- Action Item: We should enforce that traffic is only accepted from UCP.
- UCP Action Item: We should enforce authentication and authorization in UCP.

> Applications RP deploys each Application and its resources in its own namespace. 

Reviewer 2: ```suggestion
Applications RP deploys each Application and its resources in its own namespace. This is necessary for administrators of the Kubernetes cluster to correctly configure RBAC and limit the scope of permissions. 
```

#### Bicep Recipe execution

Bicep recipes can provision arbitrary infrastructure resources in the cloud, and are provided by users. In order to execute a Bicep recipe, Applications RP's Recipe Engine first authenticates with a OCI compliant registry and then  downloads the recipe from it. RP uses the stored Azure credentials to authenticate with the registry. RP also supports federated credentials. When enabled, RP uses this to authenticate with the registry. Whenever possible, federated identies should be preferred and used for robust security. After retrieving the Bicep recipe, the RP requests UCP to deploy it.

#### Terraform Recipe execution

Terraform recipes are download from internet too. We support public and private modules as well as different kinds of authentication.

Terraform downloads providers from the internet onto an empty directory `/terraform` which is mounted into the applications RP pod and executes them. This means that these providers have access to the network, filesystem, and environment variables, as well as the running memory of Applications RP. Terraform communicates with AWS as needed to deploy resources.

#### Data Serialization / Formats

> We use custom parsers to parse Radius-related resource IDs and do not use any other custom parsers. The custom parser that parses Radius resource IDs has its own security mechanisms that don't accept anything other than a Radius resource ID. 

Reviewer 2: > The custom parser that parses Radius resource IDs has its own security mechanisms that don't accept anything other than a Radius resource ID.

Does it?

I think this could be simplified to say something like.

> Radius resource ids are a custom string format that can be parsed from untrusted data. The parser is a shared component.

Beyond that is there anything else that needs to be said?

Today, the RP assumes that incoming requests are from a trusted user who is authorized to make the request. It should enforce this by checking that the requests are made by UCP. UCP in turn should accept requests from authorized and authenticated users and clients such as `rad cli`. Once we have Radius RBAC support, we should configure UCP so that it accepts requests only from authorized and authenticated users.

Applications RP deploys each Application and its resources in its own namespace. This is necessary for administrators of the Kubernetes cluster to correctly configure RBAC and limit the scope of permissions. 

## Assumptions 

This threat model assumes that:

1. The Radius installation is not tampered with.
2. The Kubernetes cluster that Radius is installed on is not compromised.
3. It is the responsibility of the Kubernetes cluster to authenticate users. Administrators and users with sufficient privileges can perform their required tasks. Radius cannot prevent actions taken by an administrator.
4. Radius stores and queues used by Applications RP are not compromised. 

> 

Reviewer 2: Let's add one more assumption here about UCP. Something like:

> HTTP requests forwarded to Applications RP from UCP originated from an authorized user. HTTP requests or other network traffic that has not been forwarded by UCP are considered unauthorized.

This document provides a threat model for the Radius Applications RP component. It identifies potential security threats to this critical part of Radius and suggests possible mitigations. The document includes an analysis of the system, its assets, identified threats, and recommended security measures to protect the system.

## Terms and Definitions

| Term                  | Definition                                                                                                                                                                                  |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| RP                 | Resource Provider    |
| UCP                 | Universal Control Plane for Radius   |
| DE | Deployment Engine |

## System Description

Applications RP is a Radius control-plane microservice that acts as resource provider for application and its resources. When users deploy an application using Radius, Applications RP manages the lifecycle of the created resources on the user's behalf. 

> - Deploying an application may launch the user application's code on the same cluster as Radius, or a different cluster. 

Reviewer: How do we achieve this? By using workspaces?

Author: yes, workspace has kube context as an entry.

> - Deploying application may create resources in the cloud using Recipes. These *recipes* are bicep or terraform code, and are responsible for creating   application infrastructure components like databases.

Reviewer: Reformatting the file is needed. There are double spaces.

> - As a result, Applications RP has access to the user's cloud credentials and can manage the user's cloud resources.

Reviewer 3: Pls correct me if I'm wrong but does applications-rp have to make a call into ucp to get credentials?

Author: We use the fetch method, like we clarified in meeting.

> Resource providers (including Applications RP) communicate over HTTP and manage the lifecycle of resources. See the [architecture documentation](https://docs.radapp.io/concepts/technical/architecture/) for more context.  Users and clients cannot directly communicate with Applications RP. They instead communicate with UCP. UCP forwards relevant requests to Applications RP. 

Reviewer 4: Do we secure the http connection?

The RP has a datastore for storing Radius data, message queue for processing asynchronous requests, and a secret Store for storing sensitive information such as certificates. All these are configurable components and support multiple implementations.  

### Architecture

> The RP consists of four types of resource providers for managing various types of resources in an application. `Applications.Core` resource provider manages core application resources such as application, environment, container and gateways. `Applications.Dapr` resource provider manages all dapr resources that are deployed as part of application. `Applications.Datastore` resource provider supports provisioning SQL database, Mongo DB and Redis Cache.

Reviewer 5: Just confirming: the Applications RP contains RPs like `Applications.Core`. In other words, we nest RPs inside of RPs.

`Applications.Messaging` resources provider manages queues such as Rabbit MQ.

Applications RP has a key sub component `Recipe Engine` to execute `recipes`. 

> `Recipes` are Bicep or Terraform code supplied by user that is used to deploy infrastructure components on Azure and AWS. The Bicep recipes are fetched from OCI compliant registries. Terraform recipes are public modules and fetched from internet too. 

Reviewer: Is it just on Azure or AWS? What about Kubernetes?

Author: I will update the text to include k8s resourecs as well.

In order to execute Terraform recipes, Applications RP installs latest Terraform. It also mounts an empty directory `/terraform` into Applications RP pod. It uses this directory for executing terraform recipes using the installed Terraform. The output resources generated from terraform module are converted to Radius output resources and stored in our datastore. 

In order to deploy bicep recipes, Applications RP sends a request to UCP, which in turn forwards it to Deployment Engine. 

The RP uses a queue to process requests asyncronously. Information about resources that are deployed / being deployed is stored in a datastore. 

> Sample high level flow:

Reviewer 5: This section is really helpful. 🚀

> Let us consider a bicep definition of application which has a container and a SQL Server DB. The container code queries the SQL Server DB. 

Reviewer: We should specify that the DB is being deployed via a Recipe.

1. Request to deploy comes from cli to UCP
2. UCP sends the deploy request to Deployment Engine
3. UCP gets container creation request from Deployment Engine and forwards it to Applications RP
4. As part of container creation, Applications RP creates the below entities and stores information about them in its datastore:
   1. Kubernetes deployment object responsible  managing for the pods of this container
   2. Kubernetes service account which provides identity for the pod
   3. Kubernetes role which defines the accesses this pod can have
   4. Kubernetes role binding which binds the role to the service account
   5. Kubernetes service if the container has ports. 

>    Since the pod has to communicates with an azure resource (SQLServer DB), Applications RP also creates a managed identity, assigns appropriate roles so that it can query the DB. It is able to do this since it has access to user's Azure credentials. 

Reviewer: Is this item number 6?

Author: cmd+shift+F

5. UCP gets SQL Server DB creation request from Deployment Engine 
6. UCP forwards the request to create SQL server DB to Applications RP
7. Applications RP communicates with OCI registry, downloads the bicep recipe for creating SQLServer DB. 

> 8.  RP then sends a request to UCP for deploying the bicep recipe and also stores information about it in its datastore.

Reviewer 5: ```suggestion
8.  The Applications RP then sends a request to UCP for deploying the bicep recipe and also stores information about it in its datastore.
```

> 9.  UCP requests DE to deploy the SQL Server DB.

Reviewer: Formatting looks off.

Author: fixed

Below is a high level overview of various key subcomponents in Applications RP
![Applications RP](2024-10-applications-rp-threat-model/apprp.png) 

### Implementation Details

#### Use of Cryptography

1. **Computing the Hash for Terraform Backend Configuration**: [Link to code](https://github.com/radius-project/radius/blob/main/pkg/recipes/terraform/config/backends/kubernetes.go#L110).

>    1. **Purpose**: The purpose of computing the hash is to compute a key for kubernetes secret which has the etrraform backend config

Reviewer 3: nit: terraform

   2. **Library**: The library used to calculate the hash of the deployment configuration is the crypto library, which is one of the standard libraries of Go: [Link to library](https://pkg.go.dev/crypto@go1.23.1).
   3. **Type**: [SHA1](https://www.rfc-editor.org/rfc/rfc3174.html). Note: "SHA-1 is cryptographically broken and should not be used for secure applications." [Link to warning](https://pkg.go.dev/crypto/sha1@go1.23.1). This is used as an optimization for detecting changes, not as a security protection.

2. Applications RP also uses UCP's queue and store libraries. These libraries use Cryptography and are detailed in UCP threat model. [link to be added]

#### Storage of secrets

> Applications RP has access to sensitive information related to the application resource it manages as well as the cloud credentials it requires for managing cloud resources on Azure and AWS. 

Reviewer: We should provide links to where and how we store the secrets.

Author: add link to ucp threat model.

##### Managing cloud credentails

Applications RP requires AWS and Azure credentials for accessing and managing resources in cloud. It fetches credentials using UCP Secret Provider library. 
Credentials are not available for retrieval through API. The RP also supports [federated identity](https://docs.radapp.io/guides/operations/providers/overview) for both Azure and AWS. Unless there is a limitation that prevents using federated identity, users should prefer using this since it removes the need to store secrets. 

Using these credentials, Applications RP can create and manage other resources in AWS and Azure. 

> #### Managing secrets for applications

Reviewer 5: ```suggestion
##### Managing secrets for applications
```

> Applications RP provides a secret store which can be used to store sensitive information such as TLS certificate and private keys. It uses kubernetes secrets to implement this secret store. In most cases, sensitive application data can be stored in this secret store and consumed by the application's other components such as a container. 

Reviewer: Is it just Kubernetes secrets implementation that is available?

Author: At present, yes.

> ###### Managing secrets for datastores

Reviewer 5: ```suggestion
##### Managing secrets for datastores
```

#### Data Serialization / Formats

We use custom parsers to parse Radius-related resource IDs and do not use any other custom parsers. Radius resource ids are a custom string format that can be parsed from untrusted data. The parser is a shared component.

We also use json for data interchange between Applications RP and other services. We use golang standard libraries to parse json. 

## Clients

**UCP** UCP is the client of Applications RP.  It forwards all requests related to Applications and several of Application resources deployment to Applications RP.

## Trust Boundaries

We have a few different trust boundaries for the Controller component:

- **Kubernetes Cluster**: The overall environment where the Applications RP  operates and receives requests from the UCP.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads.

The Applications RP component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCP also resides within the same namespace.

The Kubernetes API Server, with which Applications RP interacts, runs in the `kube-system` namespace within the cluster.

> Today, the RP assumes that incoming requests are from a trusted user who is authorized to make the request. It should enforce this by checking that the requests are made by UCP. UCP in turn should accept requests from authorized and authenticated users and clients such as `rad cli`. Once we have Radius RBAC support, we should configure UCP so that it accepts requests only from authorized and authenticated users.

Reviewer: Can a client directly call Applications RP?

Author: if it impersonates UCP, then it can.

Applications RP deploys each Application and its resources in its own namespace. This is necessary for administrators of the Kubernetes cluster to correctly configure RBAC and limit the scope of permissions. 

## Assumptions 

This threat model assumes that:

1. The Radius installation is not tampered with.
2. The Kubernetes cluster that Radius is installed on is not compromised.
3. It is the responsibility of the Kubernetes cluster to authenticate users. Administrators and users with sufficient privileges can perform their required tasks. Radius cannot prevent actions taken by an administrator.
4. Radius stores and queues used by Applications RP are not compromised. 

> HTTP requests forwarded to Applications RP from UCP originated from an authorized user. HTTP requests or other network traffic that has not been forwarded by UCP are considered unauthorized.

Reviewer: ```suggestion
5. HTTP requests forwarded to Applications RP from UCP originated from an authorized user. HTTP requests or other network traffic that has not been forwarded by UCP are considered unauthorized.
```

**Description:** 
If a malicious actor can impersonate Applications RP, requests from UCP will be received by the malicious actor. The malicious actor can also send requests to UCP and access cloud credentials.

**Impact:** 
All data sent from UCP to Applications RP will be available to the malicious actor, such as payloads of resources. These payloads can disclose information about Application resources. 
The impersonator would also be able to retreive credentials through UCP. The credentials can be used to create and consume Azure and AWS resources. The impersonator could also update Environment recipe links to use an outdated/ vulnerable version of the resource.

**Mitigations:**

1. Impersonating Applications RP, tampering with the applications rp code and configuration would require access to modify the `radius-system` namespace. Our threat model assumes that the operator has limited access to the `radius-system` namespace using Kubernetes' existing RBAC mechanism.

2. We should add support in Radius to use mTLS as communication protocol between UCP and Applications RP.

**Status:** 

> 1. Active. All mitigations which make use of kubernetes RBAC are currently active. Operators are expected to secure their cluster and limit access to the `radius-system` namespace. 

Reviewer: What about Radius RBAC? Should we talk about that too here?

> 2. Pending. 

Reviewer 5: What is pending?

Author: like we dicussed in meeting the second mitigation is pending. added full statement to clarify.

#### Malicious user can make arbitrary requests to Applications RP API. 

**Impact**

This can result in unintended application resource changes.

**Mitigation**

> Radius should add validation to make sure the incoming request is from UCP. 

Reviewer 5: Is there also an existing mitigation that is the same as the above threat? "Impersonating Applications RP, tampering with the applications rp code and configuration would require access to modify the radius-system namespace. Our threat model assumes that the operator has limited access to the radius-system namespace using Kubernetes' existing RBAC mechanism."

Reviewer 2: This is a different threat. The first one is about information disclosure.

**Status**

Pending

> #### Sniffing the communication between Applications RP  and Kubernetes API Server / UCP could cause information disclosure 

Reviewer: ```suggestion
#### Sniffing the communication between Applications RP and Kubernetes API Server / UCP could cause information disclosure 
```

**Description:** 
If a malicious actor could sniff communication between the applications RP and the UCP, the actor could modify/replay the packets.

**Impact:** 
A malicious actor could use the information about the resources and operations in progress. They can also replay the same requests to cause a DDoS or send a modified payload to cause inadvertant changes to application.

**Mitigations**

> 1. Tampering with Application RP code/ configs would require access to modify the `radius-system` namespace. Our threat model assumes that the operator has limited access to the `radius-system` namespace using Kubernetes' existing RBAC mechanism. It would be impossible for an arbitrary user to gain access to the cluster and capture traffic.

Reviewer: Maybe not impossible but we should assume that it doesn't happen.

> 2. We should add support in Radius to use mTLS as communication protocol between UCP and Applications RP.

Reviewer 4: what's the plan for this?  all network traffic should be encrypted and all data encrypted in transit and at rest

Author: It is listed as pending action item. we plan to use dapr to bring mtls support.

**Status:** 

1. Active. All mitigations listed are currently active. Operators are expected to secure their cluster and limit access to the `radius-system` namespace.

> 2. Pending

Reviewer 5: What is pending?

Author: updated to full sentence.

#### Using recipes can cause escalation of priviledge

**Description** 
A malicious user could become the admin, and configure recipes to create / update resources without having permissions to access these resources directly.

**Impact** 
This can facilitate the attackers to create resources based on any arbitrary image. 

**Mitigations**
1. Ability to register recipes should be given to only trusted employees. This in will be possible once Radius to support RBAC. 

2. Since each application resource is deployed in the application's namespace, we can use Kubernetes RBAC to confine potential security implications to the application's namepsace.

3. Any container launched by Radius is subject to Kubernetes as the platform. Users should configure their Kubernetes cluster to ensure that pods are created only from images hosted on trusted URLs or registries.

4. Users should configure RBAC of Radius with principle of least privilege so that blast radius of any security incidents involving Radius is minimized.

> 

Reviewer: reformat

**Status**
1. Pending. 
   
2. Active. Operators are expected limit access so that an application in its own namespace cannot by affected by application in another namespace.
   
3. Active. Operators are expected to configure their Kubernetes cluster with trusted URLs or registries.
   
4. Active. Operators are expected to configure RBAC of Radius appropriately to minimize its blast radius.

> #### Using application definition to deploy malicious containers

Reviewer: Is there something that we can do for this? Would you define this as a threat?

Author: I will add the kubernets's trusted url ability to limit sources from which the images can be created

**Mitigations**
Application definition and images and recipes used should be reviewed and only authorized users should be able to create and deploy application definitions.
Radius RBAC could enable only trusted users to manage application definition  recipe configurations and deployment.

**Status**
Radius RBAC support is pending. 

#### Applications RP has the a ability to create managed identities which if misused can lead to Escalation of Privilege

**Description:** If a malicious actor uses Radius, they could deploy Azure managed identity which could provide escalated priviledges to users.

**Impact:** A malicious actor use this ability to gain inappropriate access to azure resources. 

**Mitigations:**

1. Only authorized users should be able to manage application definition and  deployment using Radius. 
   
**Status:** 

Pending. This mitigation requires RBAC support in radius.

## Open Questions

## Action Items

1. Implement validation to ensure Applications RP accepts requests only from UCP. 
2. Implement Radius RBAC 
3. Introduce fuzz testing to make sure Radius parsers are robust and not vulnerable.

> 4. Radius should use mTLS as communication protocol between UCP and Applications RP.

Reviewer: Any GitHub issues created for these?

   2. **Library**: The library used to calculate the hash of the deployment configuration is the crypto library, which is one of the standard libraries of Go: [Link to library](https://pkg.go.dev/crypto@go1.23.1).
   3. **Type**: [SHA1](https://www.rfc-editor.org/rfc/rfc3174.html). Note: "SHA-1 is cryptographically broken and should not be used for secure applications." [Link to warning](https://pkg.go.dev/crypto/sha1@go1.23.1). This is used as an optimization for detecting changes, not as a security protection.

2. Applications RP also uses UCP's queue and store libraries. These libraries use Cryptography and are detailed in UCP threat model. [link to be added]

#### Storage of secrets

Applications RP has access to sensitive information related to the application resource it manages as well as the cloud credentials it requires for managing cloud resources on Azure and AWS. 

##### Managing cloud credentials

Applications RP requires AWS and Azure credentials for accessing and managing resources in cloud. It fetches credentials using UCP Secret Provider library. 

> Credentials are not available for retrieval through API. The RP also supports [federated identity](https://docs.radapp.io/guides/operations/providers/overview) for both Azure and AWS. Unless there is a limitation that prevents using federated identity, users should prefer using this since it removes the need to store secrets. 

Reviewer 2: ```suggestion
Credentials are not available for retrieval through Radius' public API. The secret provider library relies on internals of Radius to read the data from Kubernetes. Any user of the Kubernetes cluster that has read-access to secrets in the `radius-system` namespace can read these values.

The RP also supports [federated identity](https://docs.radapp.io/guides/operations/providers/overview) for both Azure and AWS. Unless there is a limitation that prevents using federated identity, users should prefer using this since it removes the need to store secrets. 
```

#### Data Serialization / Formats

We use custom parsers to parse Radius-related resource IDs and do not use any other custom parsers. Radius resource ids are a custom string format that can be parsed from untrusted data. The parser is a shared component.

We also use json for data interchange between Applications RP and other services. We use golang standard libraries to parse json. 

## Clients

**UCP** UCP is the client of Applications RP.  It forwards all requests related to Applications and several of Application resources deployment to Applications RP.

## Trust Boundaries

We have a few different trust boundaries for the Controller component:

- **Kubernetes Cluster**: The overall environment where the Applications RP  operates and receives requests from the UCP.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads.

The Applications RP component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCP also resides within the same namespace.

The Kubernetes API Server, with which Applications RP interacts, runs in the `kube-system` namespace within the cluster.

> Today, the RP assumes that incoming requests are from a trusted user who is authorized to make the request. It should enforce this by checking that the requests are made by UCP. UCP in turn should accept requests from authorized and authenticated users and clients such as `rad cli`. Once we have Radius RBAC support, we should configure UCP so that it accepts requests only from authorized and authenticated users.

Reviewer 2: ```suggestion
Today, the RP assumes that incoming requests are from an authorized user. At the time of writing there is no authentication or encryption for requests inbound to Applications RP.

We have planned work to implement TLS encryption for requests in transit, and an authentication to enforce that only UCP may send requests to Applications RP. This threat model document will be updated once this work is complete.

Applications RP relies on UCP to implement and enforce granular authorization policies. Applications RP will not implement its own authorization system. At the time of writing there is no support in UCP for granular authorization policies. This threat model document will be updated once this work is complete.
```

Below are the key points associated with data flow:
1. Applications RP receives request to deploy resources from UCP and sends back appropriate response.
2. Depending on the resource to be deployed Applications RP takes the next step 
    1. If the request is to deploy a recipe, Applications RP first downloads the recipe from a OCI registry (for bicep) or public module (for terraform)
    2. Applications RP requests UCP to deploy bicep recipes. Or if its a Terraform recipe, the downloaded recipe is executed using the installed terraform, in the empty terraform directory mounted into the Applications RP pod. As part of this execution, Terraform integrated in RP makes AWS and Azure API calls. 
    3. If it is a kubernetes resource such as container, the RP interacts with API server to render the resource.
3. Application RP uses API server to save Radius resources and Async Operations By default, API Server is used as Radius Datastore and Queue.

## Threats

#### Impersonating Applications RP could cause information disclosure, DDoS and misuse of cloud resources.

**Description:** 

> If a malicious actor can impersonate Applications RP, requests from UCP will be received by the malicious actor. The malicious actor can also send requests to UCP and access cloud credentials.

Reviewer 2: > The malicious actor can also send requests to UCP and access cloud credentials.

If this is true, it's a different threat right? This is about intercepting traffic sent from UCP -> Applications RP.

Author: broke this into two threats.

**Impact:** 
All data sent from UCP to Applications RP will be available to the malicious actor, such as payloads of resources. These payloads can disclose information about Application resources. 

> The impersonator would also be able to retrieve credentials through UCP. The credentials can be used to create and consume Azure and AWS resources. The impersonator could also update Environment recipe links to use an outdated/ vulnerable version of the resource.

Reviewer 2: I don't get this one. UCP isn't responsible for applications or environments.

> The impersonator would also be able to retrieve credentials through UCP. The credentials can be used to create and consume Azure and AWS resources. The impersonator could also update Environment recipe links to use an outdated/ vulnerable version of the resource.

Reviewer 2: > The impersonator would also be able to retrieve credentials through UCP. The credentials can be used to create and consume Azure and AWS resources. The impersonator could also update Environment recipe links to use an outdated/ vulnerable version of the resource.

Please separate this, it's a different threat.

**Mitigations:**

1. Impersonating Applications RP, tampering with the applications rp code and configuration would require access to modify the `radius-system` namespace. Our threat model assumes that the operator has limited access to the `radius-system` namespace using Kubernetes' existing RBAC mechanism.

> 2. We should add support in Radius to use mTLS as communication protocol between UCP and Applications RP.

Reviewer 2: Please replace `mTLS` with authentication. We'll decide during design what form of authentication to use.

Author: rewrote the line in terms of authentication.

**Status:** 

1. Active. All mitigations which make use of kubernetes RBAC are currently active. Operators are expected to secure their cluster and limit access to the `radius-system` namespace. 
   
2. Pending. We should add support in Radius to use mTLS as communication protocol between UCP and Applications RP.

#### Malicious user can make arbitrary requests to Applications RP API. 

**Impact**

This can result in unintended application resource changes.

**Mitigation**

> Radius should add validation to make sure the incoming request is from UCP. 

Reviewer 2: ```suggestion
Radius should add authentication to make sure the incoming request is from UCP. 
```

**Status**

Pending

#### Sniffing the communication between Applications RP  and Kubernetes API Server / UCP could cause information disclosure 

**Description:** 

> If a malicious actor could sniff communication between the applications RP and the UCP, the actor could modify/replay the packets.

Reviewer 2: How would they do this?

Author: add some explanation.

**Impact** 
This can facilitate the attackers to create resources based on any arbitrary image. 

**Mitigations**
1. Ability to register recipes should be given to only trusted employees. This in will be possible once Radius to support RBAC. 

2. Since each application resource is deployed in the application's namespace, we can use Kubernetes RBAC to confine potential security implications to the application's namespace.

3. Any container launched by Radius is subject to Kubernetes as the platform. Users should configure their Kubernetes cluster to ensure that pods are created only from images hosted on trusted URLs or registries.

4. Users should configure RBAC of Radius with principle of least privilege so that blast radius of any security incidents involving Radius is minimized.

**Status**
1. Pending. We should implement Radius RBAC and then use the feature to restrict recipe registration to only trusted employees.
   
2. Active. Operators are expected limit access so that an application in its own namespace cannot by affected by application in another namespace.
   
3. Active. Operators are expected to configure their Kubernetes cluster with trusted URLs or registries.
   

> 4. Active. Operators are expected to configure RBAC of Radius appropriately to minimize its blast radius.

Reviewer 2: There is no authorization support in Radius today, so this isn't an active mitigation.
